gluten-core/src/main/scala/io/glutenproject/extension/columnar/TransformHintRule.scala

```diff
@@ -755,12 +768,17 @@ case class AddTransformHintRule() extends Rule[SparkPlan] {
               val limitPlan = LimitTransformer(plan.child, 0, plan.limit)
               tagged = limitPlan.doValidate()
             } else {
+              val sortExec = SortExec(plan.sortOrder, false, plan.child)
+              val pulledOutSortExec =
+                ColumnarPullOutProject.getPulledOutPlanLocally[SortExec](sortExec)
```

Comment:
`ColumnarPullOutProject` will pull out `ProjectExec` and need to verify if `ProjectExec` can be converted to native plan. If we want to decouple `ColumnarPullOutProject` from `TransformHintRule`, we can place it before `TransformHintRule`, which is also feasible. Initially, I implemented it this way, but I encountered an issue where ClickHouse's [custom agg](https://github.com/oap-project/gluten/blob/25f0d69f1205ac3a64d6c2d8bf36b5b4ae6263aa/gluten-core/src/main/scala/io/glutenproject/execution/HashAggregateExecBaseTransformer.scala#L423) https://github.com/oap-project/gluten/pull/3629#discussion_r1384889177 throws an exception when determining if post-project is needed. It may require ClickHouse's assistance to redesign the API for custom agg and not rely on throwing exceptions in `getAttrsIndexForExtensionAggregateExpr` for fallback, but instead trigger the fallback logic in `doValidationInternal`. 

We can proceed with the modifications step by step. For now, let's place the validation of `ProjectExec` within the rule itself. `ColumnarPullOutProject` will only validate `ProjectExec`. I understand that there are similar codes in [other places](https://github.com/oap-project/gluten/blob/25f0d69f1205ac3a64d6c2d8bf36b5b4ae6263aa/gluten-core/src/main/scala/io/glutenproject/execution/ScanTransformerFactory.scala#L119) in Gluten that tag hints, and we can handle them together later.

Reply:
> I understand that there are similar codes in [other places](https://github.com/oap-project/gluten/blob/25f0d69f1205ac3a64d6c2d8bf36b5b4ae6263aa/gluten-core/src/main/scala/io/glutenproject/execution/ScanTransformerFactory.scala#L119) in Gluten that tag hints, and we can handle them together later.

Actually I think we may allow adding validation tags to the original physical plan in rules other than `AddTransformHintRule` as of now, just as if multiple rules are doing validation together. However I am not sure whether the code you mentioned above can be a good case so it might still be needed to optimize, I'll take a look as well.

> I encountered an issue where ClickHouse's [custom agg](https://github.com/oap-project/gluten/blob/25f0d69f1205ac3a64d6c2d8bf36b5b4ae6263aa/gluten-core/src/main/scala/io/glutenproject/execution/HashAggregateExecBaseTransformer.scala#L423) https://github.com/oap-project/gluten/pull/3629#discussion_r1384889177 throws an exception when determining if post-project is needed.

This is interesting... So my feeling is we may have to re-think how should we handle backend-specific pre/post project creation code when doing the refactor. Say, if backend A has some specialized conditions to decide whether a project should be pulled out from a plan node, the we'd provide extensibility to have it customized?

Also in the patch, code of the new feature is currently located into several places, including logical optimization, validation (transform hint), and physical optimization (the actual pulling logic). So I slightly feel that the complexity added to Gluten is a little bit higher than what we expected? Do we have chance to reduce? At the same time am just more worried about coupling of the rules in this PR. Do you think we can add some new methods to backend API to deal with the CH Agg issue you mentioned?

Reply:
The part about "custom agg" may not have been expressed clearly. Currently, both CH and Velox require a "post project" process. The logic is the same, which is to convert the output on the native side to a consistent output for Spark. What I mean is that CH throws an exception when retrieving the native output, which requires validation before pulling it out. It doesn't mean that CH "post project" process is different from Velox's and has custom requirements. In fact, it is about the `getAttrForAggregateExprs` method that retrieves the actual output of the aggregation. Based on this output, a "post-project" is constructed. Velox also has validation logic that throws exceptions. I also hope to include this logic in the `doValidationInternal` method like [this](https://github.com/oap-project/gluten/pull/4459). CH has custom aggregation requirements, and validation is also performed when retrieving the output for custom agg. For example, the `CustomSum` only supports Final [code](https://github.com/oap-project/gluten/blob/19a863c0d3781034b4def57be899e5bd4a8825b6/backends-clickhouse/src/test/scala/io/glutenproject/execution/extension/CustomAggExpressionTransformer.scala#L49). Since I am not familiar with the specific logic of other custom agg in CH, I cannot move the validation logic into `doValidationInternal` like in Velox. It may require CH's developers to redesign this part. However, this part is not essential and can be improved in future development.